File: solutions/python/day2_step3_create_amm.py
"""
Day 2 Step 3 Solution: Create AMM
===================================
Create an AMM pool with a custom token and XRP on XRPL Testnet.
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import AMMCreate
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.models.requests import AMMInfo
from xrpl.models.currencies import IssuedCurrency, XRP
from xrpl.transaction import submit_and_wait
from xrpl.utils import xrp_to_drops

TESTNET_URL = 'wss://s.altnet.rippletest.net:51233'

# Load wallets
with open('../../wallets.json') as f:
    data = json.load(f)
    wallets = data['wallets']

issuer = Wallet.from_seed(wallets[0]['seed'])
currency_code = 'TST'

print(f"Issuer: {issuer.address}")

with WebsocketClient(TESTNET_URL) as client:

    # Create AMM pool
    amm_tx = AMMCreate(
        account=issuer.address,
        amount=IssuedCurrencyAmount(
            currency=currency_code,
            issuer=issuer.address,
            value='100'
        ),
        amount2=xrp_to_drops(10),
        trading_fee=500
    )
    amm_result = submit_and_wait(amm_tx, client, issuer)
    print(f"AMMCreate: {amm_result.result['meta']['TransactionResult']}")

    # Query AMM info
    amm_info = client.request(AMMInfo(
        asset=IssuedCurrency(currency=currency_code, issuer=issuer.address),
        asset2=XRP()
    ))
    print(f"AMM Info: {amm_info.result}")

# Explorer: https://testnet.xrpl.org/accounts/{address}
print(f"\nExplorer: https://testnet.xrpl.org/accounts/{issuer.address}")

File: solutions/js/day2-step3-create-amm.js
// ============================================================
// Day 2 - Step 3 Solution: Create an AMM Pool
// Learn: How to create an Automated Market Maker (AMM) pool
// that pairs your custom token with XRP.
//
// PREREQUISITE: Run step1 (Default Ripple) and step2 (issue
// tokens) first! The issuer needs tokens to deposit into the pool.
//
// AMM Concept: An AMM pool holds two assets and allows anyone
// to swap between them. The TradingFee (in basis points) goes
// to liquidity providers.
// ============================================================

const xrpl = require('xrpl');
const fs = require('fs');

const TESTNET_URL = 'wss://s.altnet.rippletest.net:51233';

async function main() {
  const { wallets } = JSON.parse(fs.readFileSync('../../wallets.json', 'utf-8'));
  const issuerWallet = xrpl.Wallet.fromSeed(wallets[0].seed);

  const currencyCode = 'ACD';

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();
  console.log('Connected to XRPL Testnet');

  // Create AMM
  const ammCreateTx = {
    TransactionType: 'AMMCreate',
    Account: issuerWallet.address,
    Amount: {
      currency: currencyCode,
      issuer: issuerWallet.address,
      value: '100'
    },
    Amount2: xrpl.xrpToDrops('10'),
    TradingFee: 500
  };
  const result = await client.submitAndWait(ammCreateTx, { wallet: issuerWallet });
  console.log('AMM Pool created!', result.result.hash);

  // Query AMM info
  const ammInfo = await client.request({
    command: 'amm_info',
    asset: {
      currency: currencyCode,
      issuer: issuerWallet.address
    },
    asset2: { currency: 'XRP' }
  });
  console.log(JSON.stringify(ammInfo.result, null, 2));

  console.log(`\nExplorer: https://testnet.xrpl.org/accounts/${issuerWallet.address}`);
  await client.disconnect();
}

main().catch(console.error);
